presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveWriteUtils.java

```diff
@@ -556,9 +565,43 @@ public static Path createTemporaryPath(ConnectorSession session, HdfsContext con
 
         createDirectory(context, hdfsEnvironment, temporaryPath);
 
+        if (hdfsImpersonationEnabled) {
+            setDirectoryOwner(context, hdfsEnvironment, temporaryPath, targetPath);
+        }
+
         return temporaryPath;
     }
 
+    public static void setDirectoryOwner(HdfsContext context, HdfsEnvironment hdfsEnvironment, Path path, Path targetPath)
+    {
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, path);
+            FileStatus fileStatus;
+            if (!fileSystem.exists(targetPath)) {
+                // For new table
+                Path parent = targetPath.getParent();
+                if (!fileSystem.exists(parent)) {
+                    return;
+                }
+                fileStatus = fileSystem.getFileStatus(parent);
+            }
+            else {
+                // For existing table
+                fileStatus = fileSystem.getFileStatus(targetPath);
+            }
+            String owner = fileStatus.getOwner();
+            String group = fileStatus.getGroup();
+            fileSystem.setOwner(path, owner, group);
+        }
+        catch (IOException e) {
+            // Since there may be nonstandard warehouse, a newly created table's location could not be a
+            // user-unique location, for example /user/hive/warehouse/xxx. In this case, it's not possible to
+            // set owner/group as the table location's parent. In other words, setOwner operation is complement
+            // for well-built warehouse.
+            log.warn(e, "Unable to set owner of %s, it is probably due to nonstandard warehouse location", path);
```

Comment:
@findepi I still think we shouldn't fail the query when setOwner operation fail. As the comment describe, a newly created table with nonstandard location could cause setOwner fail, but it shouldn't be a fatal error.